lib/galaxy/tool_util/verify/test_data.py

```diff
@@ -122,3 +151,59 @@ def execute(self, cmd):
                 "stderr": stderr,
             }
             print(UPDATE_FAILED_TEMPLATE.substitute(**kwds))
+
+
+class RemoteLocationDataResolver(FileDataResolver):
+    def __init__(self, environ):
+        self.fetch_data = asbool(environ.get("GALAXY_TEST_FETCH_DATA", True))
+        repo_cache = environ.get("GALAXY_TEST_DATA_REPO_CACHE", "test-data-cache")
+        repo_path = os.path.join(repo_cache, "from-location")
+        super().__init__(repo_path)
+
+    def exists(self, filename: str, context: Optional[TestDataContext] = None):
+        exists_now = super().exists(filename, context)
+        if exists_now or not self.fetch_data or context is None:
+            return exists_now
+        self._try_download_from_location(filename, context)
+        exists_now = super().exists(filename, context)
+        if exists_now:
+            self._verify_checksum(filename, context)
```

Comment:
Right now, if we specify a checksum, it is validated only after the file is downloaded for the first time.
To avoid caching a corrupted file:
1 - If the file already exists, should we validate the checksum every time we check for existence? And redownload it if the check fails?
2 - Or should we delete the downloaded file right after the checksum fails?

I'm slightly inclined to option 2...